blockchain/src/blockchain/mod.rs

```diff
@@ -10,3 +13,25 @@ pub mod slots;
 pub mod verify;
 pub mod wrappers;
 pub mod zkp_sync;
+
+/// A post-validation hook that can be registered with the blockchain to run after block validation.
+/// For `extend` and `rebranch` operations, this is called before a block is committed to the database.
+/// It thus provides a timing advantage.
+pub trait PostValidationHook {
+    /// Run the pre-commit hook.
+    fn post_validation(&self, block: Block, push_result: Result<PushResult, PushError>);
```

Comment:
I think it would be nice to be able to do 
```suggestion
    fn post_validation(&self, block: &Block, push_result: Result<&PushResult, &PushError>);
```
here.

I realize that to publish this will have to be cloned within, but it will then not be cloned to do validate message, where it is unnecessary. Also the result types are not necessary to be consumed here either and can just be passed as reference.